File: greedy/activity-selection/activity-selection.cpp
// Greedy: Activity Selection (maximum non-overlapping activities)
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Activity {
    int start, end;
};

int activitySelection(vector<Activity> activities) {
    sort(activities.begin(), activities.end(), [](const Activity& a, const Activity& b) {
        return a.end < b.end;
    });

    int count = 1;
    int lastEnd = activities[0].end;

    for (size_t i = 1; i < activities.size(); i++) {
        if (activities[i].start >= lastEnd) {
            count++;
            lastEnd = activities[i].end;
        }
    }
    return count;
}

int main() {
    vector<Activity> activities = {{1, 3}, {2, 5}, {4, 6}, {6, 7}, {5, 9}, {8, 9}};

    cout << "Max activities: " << activitySelection(activities) << endl;

    return 0;
}

File: greedy/activity-selection/activity-selection.java
// Greedy: Activity Selection (maximum non-overlapping activities)
import java.util.Arrays;

class ActivitySelection {

    static int activitySelection(int[][] activities) {
        Arrays.sort(activities, (a, b) -> a[1] - b[1]);

        int count = 1;
        int lastEnd = activities[0][1];

        for (int i = 1; i < activities.length; i++) {
            if (activities[i][0] >= lastEnd) {
                count++;
                lastEnd = activities[i][1];
            }
        }
        return count;
    }

    public static void main(String[] args) {
        int[][] activities = {
            { 1, 3 }, { 2, 5 }, { 4, 6 }, { 6, 7 }, { 5, 9 }, { 8, 9 }
        };

        System.out.println("Max activities: " + activitySelection(activities));
    }
}
